### ISSUE TITLE
kde standard error and pointwise confidence intervals

### ISSUE THREAD

**Issue by josef-pkt** [statsmodels/statsmodels#1244]
based on Bruce Hansen's notes this should be easy to add (for fixed bandwidth)
http://www.ssc.wisc.edu/~bhansen/718/NonParametrics1.pdf

They might not be very good, but better than nothing.

A bit better but trickier: inverting an asymptotic t-test for confidence intervals.
non-trivial because variance depends on alternative (like in proportions tests), nonlinear rootfinding.

**Comment by cdeil** [comment]
@josef-pkt – I'm looking for a method to compute an "error band" for a KDE (don't know yet which method exactly I want).

With Google I found this in the sandbox:
https://github.com/statsmodels/statsmodels/blob/40bd7e077a8a1d73078efecd9a8f3521032e0350/statsmodels/sandbox/nonparametric/kernels.py#L233

Is there something in statsmodels I can use?

I'm thinking about implementing my own method using bootstrap as described here:
http://www.hep.caltech.edu/~NarskyPorter/Excerpts/Pages100_102.pdf
Does is exist in statsmodels or would this be a welcome addition?

**Comment by josef-pkt** [comment]
@cdeil I'm a bit out of this topic right now, and only vaguely remember this.
@PierreBdR @Padarn  Do you have more information or code?

Some confidence intervals are compared with Stata (but maybe for kernel regression) but are only in the same neighborhood because Stata uses a **pilot** estimate for constructing standard errors and confidence intervals.

Some bandwidth choice methods are based on leave one out cross-validation, but AFAIR (I'm not usre) there is no bootstrap.
For bootstrapping I would think that resampling from the gridded values for the fft version would be much faster for large samples.

**Comment by Padarn** [comment]
I have some code for doing a non-pointwise confidence interval somewhere, but I think the bootstrap method above would make a nice addition.

In the current implementation I don't think the raw gridded values (the histogram essentially) are stored, but this might be different in @PierreBdR implementation. This certainly sounds like the way to go to me.

**Comment by PierreBdR** [comment]
@cdeil In the past, I have used with success bootstrapping (see http://pyqt-fit.readthedocs.org/en/latest/KDE_tut.html ). Bootstrapping is really easy to implement for KDE, and really fast too. If this is useful, I can add a confidence interval estimator using bootstrapping to the kernel_methods module.

**Comment by PierreBdR** [comment]
@Padarn Sorry for the late answer. As I understand it (and used it), the bootstrapping relies on re-sampling the original data, with repeats.  How would you (correctly) use the histogram for this? I can imagine this is doable, but if you know of a working solution, I would be really interested in it!

**Comment by cdeil** [comment]
@PierreBdR  +1 to add a confidence interval estimator using bootstrapping!

Side question: I have data which is roughtly two Gaussians centered on 0, one very narrow with ~ 60 events and one very broad with ~ 10 events. The problem I have is that I'd like to show a KDE and an error band to indicate what are Poisson stats fluctuations. For the central part, I think bootstrapping gives a nice error band, but for the outer pars with just a few events, for most locations there's zero or one events, and bootstrapping doesn't give a meaningful error band.
The only solution I can think of is fitting a model (e.g. a double-Gauss) and drawing an error band from that model fit, propagating the parameter uncertainties.
Is there a non-parametric way to get an error band?

**Comment by josef-pkt** [comment]
@PierreBdR  If you use the original data and fft kde, then the first step would regrid the sampled bootstrap data. Since the original data points are the same as before, they would be assigned to the same grid points as initially, so this step is completely redundant in the bootstrap.

I don't know what's the fastest to resample from the gridded points or histogram. Standard resampling could be done with a full array of data points with values on the grid or on an integer index grid. I guess resampling from the histogram itself would be a binomial sample, but I never used random binomial for large arrays
`np.random.binomial(2000, np.ones(100) / 100., size=(2, 100))` looks fast

**Comment by PierreBdR** [comment]
@josef-pkt I agree this would be simpler for normal histogram. But for added accuracy, the binning is not a simple histogram but a linear binning, where each point may contribute to 2^d bins in the dataset is n dimensional (it was already done this way in the ``nonparametric`` module). So although it might very well be possible, I don't see an obvious way to simulate the re-sampling based on the gridded values only.

**Comment by PierreBdR** [comment]
@cdeil For your side question, what do you mean by "doesn't give a meaningful error band"? If you have few samples on a broad peak, I would expect the error band to be quite large! But if you have good reasons to believe you have a mix of two Gaussian distributions, you should probably use a parametric method, which will allow you to get estimates on the various parameters.

**Comment by josef-pkt** [comment]
@PierreBdR  You're right, then I don't see a way to exactly replicate the resampling from the original data, although it might be negligible in large samples with fine grid for which we use the fft/binned version.
I didn't remember the weighted allocation when the gridded data is constructed, or I never paid enough attention to this to understand the implication. This introduces an additional smoothing compared to the  plain 0-1 gridding.

Related question after a short google search: Is it recommended to undersmooth for bootstrap confidence intervals to reduce the bias?

**Comment by cdeil** [comment]
I have regions in my KDE estimate with PDF=0, i.e. no events if I choose a bandwidth that works well for the central part of my distribution.
And then any bootstrapped KDE will also have PDF=0, i.e. the "error band" has zero width there.
What I want is a band that starts at 0 and goes to some small value.

I guess I'd need some fancy adaptive-bandwith KDE to get non-zero PDF in those regions?
Are there standard KDE methods for that with readily available Python codes?
If no, I'll just pick some arbitrary model that is simple and roughly fits and fit the distribution.

**Comment by PierreBdR** [comment]
@cdeil I haven't looked too much into the estimation of variable bandwidths. But if you have a way of doing so, the ``kernel_methods`` module can estimate the KDE with them. You can to set the ``adjust`` property to an 1D array with one value per sample. Each value is the multiplicative factor used the compute the bandwidth for this sample from the default bandwidth.

**Comment by PierreBdR** [comment]
@josef-pkt I am not an expert in bootstrapping, but as far as I understand it you have to be a bit careful with its interpretation, in particular for non-parametric methods. The way I see it, this will give you an estimate on how much variability you should observe **with your method**. As for your question, I have never seen this being recommended, and I certainly wouldn't, as this seems like decreasing variability by increasing the bias, which will not be seen by the bootstrapping.

That being said, I don't see linear binning so much as a smoothing, but more as a way to reduce the noise introduced by the binning step. Another way to achieve this is to increase the number of bins in the histogram.

**Comment by PierreBdR** [comment]
@cdeil Ok, I added a bootstrap module to the ``kernel_methods`` one. I still need to add proper tests, but the notebooks ``KDE1D`` and ``KDEnD`` in the tests folder of ``kernel_methods`` show how to use this.

**Comment by josef-pkt** [comment]
I'm parking here some references that I found interesting while reading about related questions in the last few days, mostly Stata (user packages) related

http://fmwww.bc.edu/RePEc/bocode/k/kdens.pdf "Univariate kernel density estimation" looks like a good collection of formulas but short on text

http://www.stata-journal.com/sjpdf.html?articlenum=st0064 "Confidence intervals for kernel density
estimation" 
includes argument and default for undersmoothing for the confidence intervals

Cameron Trivedi on undersmoothing for confidence intervals
https://books.google.fr/books?id=Zf0gCwxC9ocC&pg=PA305&lpg=PA305&dq=confidence+interval+for+kernel+density&source=bl&ots=CZ5bpN6FqV&sig=aVF91YIbXLwCasM8jzLrMzK_6RU&hl=en&sa=X&ved=0CDAQ6AEwAzgKahUKEwiS0t3Nte7GAhXGaRQKHYYkCRM#v=onepage&q=confidence%20interval%20for%20kernel%20density&f=false

there are also several articles about implementing adaptive, variable bandwidth kernel density estimator (univariate) e.g. http://econpapers.repec.org/article/tsjstataj/v_3a3_3ay_3a2003_3ai_3a2_3ap_3a148-156.htm

and a few more related older articles in Stata journals

**Comment by josef-pkt** [comment]
another reference from Statistical Machine learning http://www.stat.cmu.edu/~larry/=sml/densityestimation.pdf

Essentially, confidence intervals that are based on an optimal bandwidth for density estimation ignore the bias and will not have the correct coverage probability. This also applies to bootstrap. To get confidence intervals with correct coverage we have to correct for bias, or remove bias so it's small relative to the variance of the estimator.

**Comment by PierreBdR** [comment]
@josef-pkt Thanks a lot for digging out these references! I will look at them and see how I can integrate them into the ``kernel_method`` package.

**Comment by PierreBdR** [comment]
Hi and sorry for the long wait. There is one thing I find deeply unsatisfying about these papers: they (rightly) say that bootstrapping requires a non-biased method to work well. They also (rightly) point to the inherent bias in the kernel methods to estimate KDE. But they fail to provide an alternative method.

I would add two references, which tackle this problem, and if nobody opposes this, I'm going to implement the method they propose:
 - http://www.jstor.org/stable/2241978
 - http://www.jdsruc.org/upload/JDS1230(2015-07-01183341).pdf

The first paper shows why you need to reduce the bandwidth, and the second provides a method to estimate the extent of this reduction. The idea will be to provide two functions: the bootstrap function will accept a new bandwidth for parameter; and I will provide a new bandwidth estimator that implements the second paper to provide a bandwidth adapted to the data. Depending on performance, the new bandwidth estimator will be the default for the bootstrap function.

**Comment by josef-pkt** [comment]
I didn't look much into KDE recently, but the same problem for confidence intervals and bias variance tradeoff shows up in non or semi parametric regression (splines and GAM)

I was reading several papers for splines and GAM this summer, the articles are quite recent and there seems to be still ongoing research on this question. In the articles the authors that their method is doing very well compared to the other ones, but I don't see clear winners and some are not easy to implement. 
My impression was that we will have to improve these over time, and keep or add several options.
https://groups.google.com/d/msg/pystatsmodels/SGwK0qBf138/FGb42ey5BQAJ

For example,  Bruce Hansen in his lecture notes (that we also used for some other improvements in nonparametrics) argues just to ignore the bias in the nonparametric regression case and stick with standard confidence intervals.

**Comment by PierreBdR** [comment]
Ok, I added the basic bootstrap function with an argument to adjust the bandwidth (by default it does nothing). I also added tests to ensure it works somewhat reasonably for the PDF estimation.

**Comment by aphearin** [comment]
I use the `statsmodels` implementation of `KDEMultivariate` to estimate the conditional PDF of some data I'm modeling. Since I want to fit my model to the KDE-estimated conditional PDFs, I need error bars on the PDF (ideally a covariance matrix). I *think* this is the concern of this Issue, but maybe I misinterpret? If so, what is the status of this issue?

**Comment by josef-pkt** [comment]
@aphearin 
I guess there is nothing for conditional pdf confidence intervals. AFAIR, I never read anything about it for the conditional pdf case. Most of the discussion and references are for the simpler univariate case, where it is already difficult to find good confidence intervals.

I don't have access right now to the Racine and Li textbook to check whether they have anything about it. The main additional difficulty for analytic, asymptotic standard errors and confidence interval would come from the property that the conditional pdf/kde is the ratio of two estimated pdf/kde, I think.

A background question: What are you doing with the conditional pdf?
One problem, I see is that the pdf is a two dimensional function (when we have two variables), so the question is for how many or which points (y, x) we want to compute confidence intervals for the conditional f(y | x).

(Aside: For similar purposes we can also use QuantileRegression with splines when the emphasis is on quantiles of the conditional distribution. We don't have `quantile sheets` yet, where many quantile functions are computed at the same time and prevent quantile crossings.)

**Comment by aphearin** [comment]
Thanks for the clarification, @josef-pkt . I may need to resort to binned methods if this is a thorny problem. 

I have an astronomical dataset that gives me a set of datapoints, each with values for the following three variables, _A_, _B_, _x_. What I'm doing is building a simple toy model for a physical quantity _x_, and I want to try and capture the observed dependence of _x_ upon _A_ and _B_. So, I have some model parameters, {p_mu}, and for each choice of {p_mu} I can predict _x_, _A_ and _B_. For the purpose of this toy model, the predictions for _A_ and _B_ are already very well-constrained, so I already know that the model successfully captures the observed P(_A_, _B_). This means that I only really want to try and successfully predict P(_x_ | _A_, _B_), evaluated at just a few control points of _A_ and _B_. So what I want to do is run an MCMC over the model parameters {p_mu}, fitting my model to a few tabulations of P(_x_ | _A_, _B_). This means I need some kind of error estimation on P(_x_ | _A_, _B_). I have been using KDE to estimate P(_x_ | _A_, _B_) from the data, but without errors, I can't compute chi^2, so I can't run an MCMC to get model posteriors. 

My last resort will simply be to manually bin P(_x_ | _A_, _B_), and get errors from bootstrap covariance. If you know of a smarter way to go about this using KDE, I'd happily pursue it.

**Comment by josef-pkt** [comment]
@aphearin I only have vague ideas about this, and no time this summer to look into it.
I opened a new issue #3741 specifically for the conditional and multivariate cases.
If you have or know about any references for this, then it would be helpful to add to the issue to get started.

(If it's just a few points of the conditioning variables, then something close to the univariate case might work. In general, I don't know if we could use conditional kernels, instead of the full Multivariate KDE, for this like in kernel regression, but using counts instead of the regressor values. "zero order" polynomial similar to having kde as special case for regression in the difference based estimators #3492 or #3380)

**Comment by aphearin** [comment]
Thanks @josef-pkt , I'll keep an eye on this issue. I'll also ask around my stats department and will post to #3741 if I learn anything useful.
